pkg/webhooks/sgx/sgx.go

```diff
@@ -51,26 +81,89 @@ func (s *SgxMutator) Handle(ctx context.Context, req admission.Request) admissio
 	}
 
 	totalEpc := int64(0)
+	epcUserCount := int32(0)
+	aesmdPresent := bool(false)
 
 	if pod.Annotations == nil {
-		pod.Annotations = map[string]string{}
+		pod.Annotations = make(map[string]string)
 	}
+
+	quoteProvider := pod.Annotations[quoteProvAnnotation]
+
 	for idx, container := range pod.Spec.Containers {
 		requestedResources, err := containers.GetRequestedResources(container, namespace)
 		if err != nil {
 			return admission.Errored(http.StatusInternalServerError, err)
 		}
-		if epcSize, ok := requestedResources[epc]; ok {
-			totalEpc += epcSize
 
-			attestation, found := pod.Annotations[provisionAnnotation]
-			if found && attestation == "yes" {
-				pod.Spec.Containers[idx].Resources.Limits[corev1.ResourceName(provision)] = resource.MustParse("1")
-				pod.Spec.Containers[idx].Resources.Requests[corev1.ResourceName(provision)] = resource.MustParse("1")
+		// the container has no sgx.intel.com/epc
+		epcSize, ok := requestedResources[epc]
+		if !ok {
+			continue
+		}
+
+		totalEpc += epcSize
+		epcUserCount += 1
+
+		// Quote Generation Modes:
+		//
+		// in-process: A container has its own quote provider library library: In this mode,
+		// the container needs a handle to /dev/sgx/provision (sgx.intel.com/provision resource).
+		// out-of-process: A container uses Intel aesmd. In this mode, the container must talk to
+		// aesmd over /var/run/aesmd/aesm.sock. aesmd can run either as a side-car or a DaemonSet
+		//
+		// Mode selection: The mode selection is done by setting sgx.intel.com/quote-provider annotation
+		// to a value that specifies the container name. If the annotation matches the container requesting
+		// SGX EPC resources, the webhook adds both /dev/sgx/provision and /dev/sgx/enclave resource requests.
+		// Without sgx.intel.com/quote-provider annotation set, the container is not able to generate quotes
+		// for its enclaves. When pods set sgx.intel.com/quote-provider: "aesmd", Intel aesmd specific volume
+		// mounts are added. In both DaemonSet and sidecar deployment scenarios for aesmd, its container name
+		// must be set to "aesmd" (TODO: make it configurable?).
```

Comment:
Maybe in the future versions we could use an annotation instead of container name (for example: `sgx.intel.com/provisioningProvider`) in case of some other component other than aesmd would implement the same interface. This would require the interface to be well specified though.

Reply:
good point. OTOH, currently this webhook is aesmd/QPL specific so any other implementation could have it's own webhook running